server/tools.go

```diff
@@ -0,0 +1,414 @@
+package server
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	gotmpl "text/template"
+
+	jsonv2 "github.com/go-json-experiment/json"
+	jsontext "github.com/go-json-experiment/json/jsontext"
+
+	"github.com/ollama/ollama/api"
+)
+
+type State int
+
+// TODO: potentially coalesce states
+const (
+	SendTokens State = iota
+	GreedyToolWithPrefix
+	GreedyToolNoPrefix
+	ForceTools
+	ToolSuffix
+	ContainsPrefix
+	PartialPrefix
+	NotPartialPrefix
+	Done
+)
+
+type ExternalState int
+
+const (
+	ToolCallFound ExternalState = iota
+	ToolCallSendPartial
+	ToolCallAccumulate
+	ToolCallSendTokens
+)
```

Comment:
ended up separating them out as there is a different representation of what action to take from a caller perspective and an internal perspective.